File: 14-TypeScript/Grey Multi-Tenant/packages/grey-matlab/tests/OptionsTest.m
classdef OptionsTest < matlab.unittest.TestCase
    %OPTIONSTEST Tests for grey.config.Options
    
    methods (Test)
        function testDefaultConstruction(testCase)
            opts = grey.config.Options('api.grey.com');
            testCase.assertEqual(opts.host, 'api.grey.com');
            testCase.assertEqual(opts.port, 443);
            testCase.assertTrue(opts.useTls);
            testCase.assertEqual(opts.timeout, 30);
        end
        
        function testFullConstruction(testCase)
            headers = struct('X_Custom', 'value');
            opts = grey.config.Options('localhost', 8080, false, 60, headers);
            testCase.assertEqual(opts.host, 'localhost');
            testCase.assertEqual(opts.port, 8080);
            testCase.assertFalse(opts.useTls);
            testCase.assertEqual(opts.timeout, 60);
            testCase.assertEqual(opts.headers.X_Custom, 'value');
        end
        
        function testLocalFactory(testCase)
            opts = grey.config.Options.local();
            testCase.assertEqual(opts.host, 'localhost');
            testCase.assertEqual(opts.port, 8080);
            testCase.assertFalse(opts.useTls);
        end
        
        function testLocalFactoryCustomPort(testCase)
            opts = grey.config.Options.local(3000);
            testCase.assertEqual(opts.port, 3000);
        end
        
        function testProductionFactory(testCase)
            opts = grey.config.Options.production('api.grey.com');
            testCase.assertEqual(opts.host, 'api.grey.com');
            testCase.assertEqual(opts.port, 443);
            testCase.assertTrue(opts.useTls);
        end
        
        function testProductionFactoryCustomPort(testCase)
            opts = grey.config.Options.production('api.grey.com', 8443);
            testCase.assertEqual(opts.port, 8443);
        end
        
        function testBaseUrlHttp(testCase)
            opts = grey.config.Options.local(8080);
            testCase.assertEqual(opts.baseUrl(), 'http://localhost:8080');
        end
        
        function testBaseUrlHttps(testCase)
            opts = grey.config.Options.production('api.grey.com');
            testCase.assertEqual(opts.baseUrl(), 'https://api.grey.com:443');
        end
        
        function testWithTimeout(testCase)
            opts = grey.config.Options.local();
            newOpts = opts.withTimeout(120);
            testCase.assertEqual(newOpts.timeout, 120);
            testCase.assertEqual(opts.timeout, 30); % Original unchanged
        end
        
        function testWithHeaders(testCase)
            opts = grey.config.Options.local();
            newOpts = opts.withHeaders(struct('Authorization', 'Bearer token'));
            testCase.assertEqual(newOpts.headers.Authorization, 'Bearer token');
            testCase.assertEmpty(fieldnames(opts.headers)); % Original unchanged
        end
    end
end

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-php/tests/OptionsTest.php
<?php

declare(strict_types=1);

namespace Grey\Sdk\Tests;

use Grey\Sdk\Config\GreyOptions;
use PHPUnit\Framework\TestCase;

final class OptionsTest extends TestCase
{
    public function testOptionsCreation(): void
    {
        $options = new GreyOptions(
            host: 'api.grey.com',
            port: 443,
            useTls: true,
            timeoutMs: 5000,
            metadata: ['X-Custom' => 'value'],
        );

        $this->assertSame('api.grey.com', $options->getHost());
        $this->assertSame(443, $options->getPort());
        $this->assertTrue($options->useTls());
        $this->assertSame(5000, $options->getTimeoutMs());
        $this->assertSame(5.0, $options->getTimeoutSeconds());
        $this->assertSame(['X-Custom' => 'value'], $options->getMetadata());
    }

    public function testOptionsBaseUrl(): void
    {
        $httpOptions = new GreyOptions('localhost', 8080, false);
        $httpsOptions = new GreyOptions('api.grey.com', 443, true);

        $this->assertSame('http://localhost:8080', $httpOptions->getBaseUrl());
        $this->assertSame('https://api.grey.com:443', $httpsOptions->getBaseUrl());
    }

    public function testLocalFactory(): void
    {
        $options = GreyOptions::local();
        
        $this->assertSame('localhost', $options->getHost());
        $this->assertSame(8080, $options->getPort());
        $this->assertFalse($options->useTls());
    }

    public function testLocalFactoryWithPort(): void
    {
        $options = GreyOptions::local(9000);
        
        $this->assertSame(9000, $options->getPort());
    }

    public function testProductionFactory(): void
    {
        $options = GreyOptions::production('api.grey.com');
        
        $this->assertSame('api.grey.com', $options->getHost());
        $this->assertSame(443, $options->getPort());
        $this->assertTrue($options->useTls());
    }

    public function testWithTimeout(): void
    {
        $options = GreyOptions::local()->withTimeout(10000);
        
        $this->assertSame(10000, $options->getTimeoutMs());
        $this->assertSame('localhost', $options->getHost()); // Original values preserved
    }

    public function testWithMetadata(): void
    {
        $options = GreyOptions::local()
            ->withMetadata(['X-First' => 'a'])
            ->withMetadata(['X-Second' => 'b']);
        
        $metadata = $options->getMetadata();
        $this->assertSame('a', $metadata['X-First']);
        $this->assertSame('b', $metadata['X-Second']);
    }

    public function testToArray(): void
    {
        $options = new GreyOptions('localhost', 8080, false, 30000, ['key' => 'value']);
        $array = $options->toArray();

        $this->assertSame('localhost', $array['host']);
        $this->assertSame(8080, $array['port']);
        $this->assertFalse($array['useTls']);
        $this->assertSame(30000, $array['timeoutMs']);
        $this->assertSame(['key' => 'value'], $array['metadata']);
    }
}
